device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/IotHubConnectionStatusChangeCallbackHandler.java

```diff
@@ -0,0 +1,95 @@
+                    case COMMUNICATION_ERROR:
+                        log.debug("### The DeviceClient is retrying based on the retry policy. Do NOT close or open the DeviceClient instance");
+                        log.debug("### The DeviceClient can still queue messages and report properties, but they won't be sent until the connection is established.");
+                        return;
+                    case CONNECTION_OK:
+                    case BAD_CREDENTIAL:
+                    case EXPIRED_SAS_TOKEN:
+                    case RETRY_EXPIRED:
+                    case CLIENT_CLOSE:
+                    default:
+                        throw new IllegalStateException(String.format("### DeviceClient cannot be %s with reason %s", status, statusChangeReason));
+                }
+            case DISCONNECTED:
+                switch (statusChangeReason) {
+                    case CLIENT_CLOSE:
+                        log.debug("### The DeviceClient has been closed gracefully. You can reopen by calling open() on this client.");
+                        return;
+                    case BAD_CREDENTIAL:
+                    case EXPIRED_SAS_TOKEN:
+                        log.debug("### The supplied credentials were invalid. Fix the input and create a new device client instance.");
+                        return;
+                    case RETRY_EXPIRED:
+                        log.debug("### The DeviceClient has been disconnected because the retry policy expired. Can be reopened by closing and then opening the instance.");
+                        if (Application.reconnectIndefinitely) {
+                            new Thread(new Runnable() {
+
+                                @Override
+                                public void run() {
+                                    clientManager.reconnect();
+                                }
+                            }).start();
+                        }
+                        return;
+                    case COMMUNICATION_ERROR:
+                        log.debug("### The DeviceClient has been disconnected due to a non-retryable exception. Inspect the throwable for details.");
+                        log.debug("### The DeviceClient can be reopened by closing and then opening the instance.");
+                        if (Application.reconnectIndefinitely) {
+                            new Thread(new Runnable() {
+
+                                @Override
+                                public void run() {
+                                    clientManager.reconnect();
+                                }
+                            }).start();
+                        }
+                        return;
+                    case CONNECTION_OK:
+                    case NO_NETWORK:
+                    default:
+                        throw new IllegalStateException(String.format("### DeviceClient cannot be %s with reason %s", status, statusChangeReason));
```

Comment:
>IllegalStateException [](start = 34, length = 21)

Don't throw an exception in this thread. Just logging is fine here

Reply:
The reason I am throwing this exception here is that we don't expect the device client to ever reach this state; if it does, the application should abort to indicate that the device client reached an unexpected state.

Reply:
The thread that this callback lives on isn't built to handle an IllegalStateException being thrown here. For example, if the reconnect thread extends this callback and you throw an exception, the transport layer will be left in a bad state.